File: plugins/geo_fencing/android/src/main/kotlin/com/example/geo_fencing/actions/MethodAction.kt
package com.example.geo_fencing.actions

import com.example.geo_fencing.actions.result.ActionResult
import com.example.geo_fencing.actions.result.FailureResult
import io.flutter.plugin.common.MethodCall

interface MethodAction {


    fun createAction(method: MethodCall): ActionResult

    fun action(method: MethodCall): ActionResult {
        return if (hasValidArguments(method)) {
            createAction(method)
        } else {
            FailureResult("required arguments not set")
        }
    }

    fun hasValidArguments(method: MethodCall): Boolean {
        return true
    }

    fun hasStringArgument(method: MethodCall, argument: String): Boolean {
        return method.argument<String>(argument)?.isNotEmpty() == true
    }

}

File: plugins/geo_fencing/ios/Classes/actions/MethodAction.swift
//Made by Vahid Rajabi

import Flutter

protocol MethodAction {
    func createAction(method: FlutterMethodCall) -> ActionResult
    func action(method: FlutterMethodCall) -> ActionResult
    func hasValidArguments(method: FlutterMethodCall) -> Bool
}

extension MethodAction {
    func action(method: FlutterMethodCall) -> ActionResult {
            if hasValidArguments(method: method) {
                return createAction(method: method)
            } else {
                return FailureResult(error: "required arguments not set")
            }
        }
        
        func hasValidArguments(method: FlutterMethodCall) -> Bool {
            return true
        }
        
        func hasStringArgument(method: FlutterMethodCall, argument: String) -> Bool {
            guard let args = method.arguments as? [String: Any],
                  let arg = args[argument] as? String,
                  !arg.isEmpty else {
                return false
            }
            return true
        }
}
